src/nvim/os/env.c

```diff
@@ -359,8 +372,37 @@ void init_homedir(void)
       }
     }
 #endif
-    homedir = xstrdup(var);
   }
+
+  // As a last resort, return the current working directory
+  if (var == NULL && os_dirname((char_u *)os_buf, MAXPATHL) == OK) {
+    var = (char *)os_buf;
+  }
+
+  set_homedir(var, MAXPATHL);
+}
+
+const char *os_homedir(void)
+{
+    size_t homedir_size = (size_t)MAXPATHL;
+    int ret = uv_os_homedir((char *)os_buf, &homedir_size);
+    if (ret == 0 && homedir_size > 0) {
+        return xstrndup((char *)os_buf, homedir_size);
```

Comment:
Is the memory allocated here freed somewhere?

Reply:
You are right, I should be freeing allocated memory.

Reply:
Actually I think my set_homedir() should catch free all the new lines added. I think homedrive and homepath need to be freed.

Reply:
Never mind set_homedir() does not free the input string. I will add that.

Reply:
Looks like freeing memory is a bigger project than expected. Somehow, when I free var (guessed homedir) at the end, I'm failing functional tests. My theory is that var sometimes is pointing at things shared by other pointers.

Reply:
What about the next change?

```C
diff --git a/src/nvim/os/env.c b/src/nvim/os/env.c
index 2769eb44b..50d1e80b2 100644
--- a/src/nvim/os/env.c
+++ b/src/nvim/os/env.c
@@ -299,6 +299,7 @@ void init_homedir(void)
   homedir = NULL;

   const char *var = os_getenv("HOME");
+  bool must_free = false;

 #ifdef WIN32
   // Typically, $HOME is not defined on Windows, unless the user has
@@ -341,6 +342,7 @@ void init_homedir(void)

   if (var == NULL) {
      var = os_homedir();
+     must_free = true;
   }

   // Default home dir is C:/
@@ -357,6 +359,7 @@ void init_homedir(void)
   // block to resolve links
   if (var == NULL) {
     var = os_homedir();
+    must_free = true;
   }

   if (var != NULL) {
@@ -380,6 +383,9 @@ void init_homedir(void)
   }

   set_homedir(var, MAXPATHL);
+  if (must_free) {
+    xfree((void *)var);
+  }
 }

 const char *os_homedir(void)
```

Reply:
I particular, os_getenv("HOME") puts ("HOME", "path") pair in a map, so if I free that, I might end up freeing the value in that map. So if os_getenv("HOME") gets called somewhere else, I might run into issues.

Reply:
> What about the next change?
> 
> ```c
> diff --git a/src/nvim/os/env.c b/src/nvim/os/env.c
> index 2769eb44b..50d1e80b2 100644
> --- a/src/nvim/os/env.c
> +++ b/src/nvim/os/env.c
> @@ -299,6 +299,7 @@ void init_homedir(void)
>    homedir = NULL;
> 
>    const char *var = os_getenv("HOME");
> +  bool must_free = false;
> 
>  #ifdef WIN32
>    // Typically, $HOME is not defined on Windows, unless the user has
> @@ -341,6 +342,7 @@ void init_homedir(void)
> 
>    if (var == NULL) {
>       var = os_homedir();
> +     must_free = true;
>    }
> 
>    // Default home dir is C:/
> @@ -357,6 +359,7 @@ void init_homedir(void)
>    // block to resolve links
>    if (var == NULL) {
>      var = os_homedir();
> +    must_free = true;
>    }
> 
>    if (var != NULL) {
> @@ -380,6 +383,9 @@ void init_homedir(void)
>    }
> 
>    set_homedir(var, MAXPATHL);
> +  if (must_free) {
> +    xfree((void *)var);
> +  }
>  }
> 
>  const char *os_homedir(void)
> ```

Nice, that's much cleaner my current thing that keeps tracks of every var change.

Reply:
My patch was incomplete. After `os_homedir()` is called, we must check `must_free` before assigning to `var` to free the memory, and change `must_free` after assignment.

It may be easier to use a new `MAXPATHL` length buffer with auto variables in the following places　and return `os_buf` in `os_homedir()`.

https://github.com/neovim/neovim/blob/b24dc5a0f236f40783e1bde21d81cdabcb8a10ad/src/nvim/os/env.c#L385-L386